v1/contrib/lambda/wrapper.go

```diff
@@ -0,0 +1,367 @@
+	typ string
+	err error
+}
+
+func (w *traceWrapper) After(result interface{}, err *typedError, endArgs ...interface{}) interface{} {
+	statusCode := 0
+	errType := ""
+	errClass := ""
+	errMsg := ""
+
+	if err != nil {
+		errType = err.typ
+		errClass = "error"
+		errMsg = err.err.Error()
+		statusCode = 500
+	}
+
+	if statusCode == 0 {
+		if gwRsp, ok := result.(events.APIGatewayProxyResponse); ok {
+			statusCode = gwRsp.StatusCode
+		} else if gwRsp, ok := result.(events.APIGatewayV2HTTPResponse); ok {
+			statusCode = gwRsp.StatusCode
+		}
+		if statusCode/100 == 5 {
+			errClass = "HTTP Error"
+			errMsg = strconv.Itoa(statusCode)
+			errType = "status"
+		}
+	}
+
+	if statusCode != 0 {
+		endArgs = append(endArgs, "Status", statusCode)
+		if statusCode/100 == 5 {
+			endArgs = append(endArgs, "ErrorClass", errClass)
+			endArgs = append(endArgs, "ErrorType", errType)
+			endArgs = append(endArgs, "ErrorMsg", errMsg)
+		}
+	}
+
+	defer w.trace.End(endArgs...)
+	if w.mdStr != "" {
+		return w.injectTraceContext(result, w.trace.ExitMetadata())
+	} else {
+		return result
+	}
+}
+
+func (w *traceWrapper) injectTraceContext(result interface{}, mdStr string) (res interface{}) {
+	defer func() {
+		if err := recover(); err != nil {
+			res = result
+		}
+	}()
+
+	if gwRsp, ok := result.(events.APIGatewayProxyResponse); ok {
+		if gwRsp.Headers != nil {
+			gwRsp.Headers[AOHTTPHeader] = mdStr
+			return result
+		}
+	} else if gwRsp, ok := result.(events.APIGatewayV2HTTPResponse); ok {
+		if gwRsp.Headers != nil {
+			gwRsp.Headers[AOHTTPHeader] = mdStr
+			return result
+		}
+	} else {
+		return result
+	}
+
+	rVal := reflect.Indirect(reflect.ValueOf(result))
+	rType := rVal.Type()
+	if rVal.Kind() != reflect.Struct {
+		return result
+	}
+
+	newResult := reflect.New(rType).Elem()
+
+	for i := 0; i < newResult.NumField(); i++ {
+		fieldVal := reflect.Indirect(newResult.Field(i))
+		fieldType := newResult.Type().Field(i)
+
+		fieldVal.Set(rVal.Field(i))
+		if fieldType.Name == "Headers" &&
+			fieldType.Type.Kind() == reflect.Map &&
+			fieldType.Type.Elem().Kind() == reflect.String &&
+			fieldType.Type.Key().Kind() == reflect.String &&
+			fieldVal.CanSet() {
+			traceContextMap := map[string]string{AOHTTPHeader: mdStr}
+			fieldVal.Set(reflect.Indirect(reflect.ValueOf(traceContextMap)))
+		}
+	}
+
+	return newResult.Interface()
+}
+
+// Wrap wraps the AWS Lambda handler and do the instrumentation. It
+// returns a new handler and can be passed into lambda.Start()
+func Wrap(handlerFunc interface{}) interface{} {
+	return HandlerWithWrapper(handlerFunc, &traceWrapper{})
+}
+
+// HandlerWithWrapper wraps a function with the wrapper and returns the wrapped function.
+func HandlerWithWrapper(handlerFunc interface{}, w Wrapper) interface{} {
+	if err := checkSignature(reflect.TypeOf(handlerFunc)); err != nil {
+		// Does not wrap an invalid handler but let lambda.Start() reject it later
+		return handlerFunc
+	}
+
+	invocationCount := 1
```

Comment:
is `HandlerWithWrapper()` only called once during cold start and subsequent calls (warm starts) only execute the anonymous function `func(ctx context.Context, msg json.RawMessage)` below?

Reply:
Yes, the anonymous function is the one that gets executed in a lambda call.